cmd/snap-update-ns/change.go

```diff
@@ -59,26 +60,139 @@ var changePerform func(*Change) ([]*Change, error)
 
 // changePerformImpl is the real implementation of Change.Perform
 func changePerformImpl(c *Change) ([]*Change, error) {
+	var changes []*Change
 	if c.Action == Mount {
+		// We may be asked to bind mount a file, bind mount a directory, mount
+		// a filesystem over a directory or create a symlink (which is abusing
+		// the "mount" concept slightly). That actual operation is performed in
+		// c.lowLevelPerform. Here we just set the stage to make that possible.
+		kind, _ := c.Entry.OptStr("x-snapd.kind")
+		path := c.Entry.Dir
+
+		// We use lstat to ensure that we don't follow a symlink in case one
+		// was set up by the snap. Note that at the time this is run, all the
+		// snap's processes are frozen.
+		fi, err := osLstat(path)
+
+		// In case we need to create something, some constants.
 		const (
 			mode = 0755
 			uid  = 0
 			gid  = 0
 		)
-		// Create target mount directory if needed.
-		if err := ensureMountPoint(c.Entry.Dir, mode, uid, gid); err != nil {
-			return nil, err
+
+		if err != nil && os.IsNotExist(err) {
+			// If the element doesn't exist we can attempt to create it.
+			// We will create the parent directory and then the final element
+			// relative to it. The traversed space may be writable so we try a
+			// to create the parent directory first.
+			switch kind {
+			case "":
+				err = secureMkdirAll(path, mode, uid, gid)
+			case "file":
+				err = secureMkfileAll(path, mode, uid, gid)
+			case "symlink":
+				// Symlinks are created later in c.lowLevelPerform but we can
+				// create the parent directory here.
+				path = filepath.Dir(c.Entry.Dir)
+				err = secureMkdirAll(path, mode, uid, gid)
+				if err == nil {
+					err = c.lowLevelPerform()
+				}
+			}
+
+			if err2, ok := err.(*ReadOnlyFsError); err2 != nil && ok {
```

Comment:
If `ok` is true, then `err2 != nil`, also `err2, _ := err(*ReadOnlyFsError); err2 != nil` is fine too

Reply:
Good point, I'll simplify this.